File: src/features/sea/learn_loop.zig
const std = @import("std");
const build_options = @import("build_options");
const wdbx = if (build_options.feat_wdbx) @import("../wdbx/mod.zig") else @import("../wdbx/stub.zig");
const ai = if (build_options.feat_ai) @import("../ai/mod.zig") else @import("../ai/stub.zig");
const router = @import("../ai/router.zig");
const evidence = @import("evidence.zig");
const query_plan = @import("query_plan.zig");
const core_memory = @import("../../core/memory.zig");

/// Configuration for one self-learning pass.
pub const LearnLoopConfig = struct {
    /// Number of prior records to recall as evidence.
    evidence_limit: usize = 5,
    /// Persist the completion (query/response vectors + block) into the store.
    persist: bool = true,
    /// Update + save the adaptive persona-router weights from this turn.
    adapt_router: bool = true,
    /// Hard cap on the augmented prompt (mirrors `evidence.MAX_PROMPT_BYTES`).
    max_prompt_bytes: usize = evidence.MAX_PROMPT_BYTES,
    /// Optional memory tracker. When set, the adaptive-weight save's transient
    /// serialize buffer is routed through a `TrackingAllocator` so weight
    /// persistence is observable. The buffer is balanced (alloc+free inside
    /// `saveWeights`), so no caller-owned/escaping allocation is tracked.
    tracker: ?*core_memory.MemoryTracker = null,
};

/// Result of `runLearnLoop`. Owns the underlying completion; `deinit` frees it.
pub const LearnLoopResult = struct {
    completion: ai.CompletionResult,
    evidence_count: usize,
    adapted: bool,
    /// The task intent inferred from the input for this turn (drives task-aware
    /// retrieval). `.general` when nothing more specific matched.
    query_task: query_plan.TaskType = .general,

    pub fn deinit(self: *LearnLoopResult, allocator: std.mem.Allocator) void {
        self.completion.deinit(allocator);
    }
};

/// Run one SEA self-learning pass:
///   1. gather evidence relevant to `input`
///   2. augment the prompt with recalled snippets
///   3. load the adaptive persona-router weights
///   4. complete (persisting per `config.persist`)
///   5. update + save the router weights from this turn (per `config.adapt_router`)
///
/// Reuses the existing `ai.completeWithStore` and `router.AdaptiveModulator`;
/// no new ML is introduced.
pub fn runLearnLoop(
    allocator: std.mem.Allocator,
    store: *wdbx.Store,
    input: []const u8,
    model: []const u8,
    config: LearnLoopConfig,
) !LearnLoopResult {
    // Infer a task-aware plan once and thread it through retrieval, so e.g. a
    // project_recall query shifts evidence weighting toward exact wording.
    const plan = query_plan.infer(input);
    var ctx = try evidence.gatherEvidenceWithPlan(allocator, store, input, config.evidence_limit, plan);
    defer ctx.deinit();
    const evidence_count = ctx.items.len;

    const augmented = try evidence.augmentPrompt(allocator, input, &ctx);
    defer allocator.free(augmented);

    var modulator = router.AdaptiveModulator.loadWeights(store);

    var completion = try ai.completeWithStoreAdaptive(allocator, store, .{
        .input = augmented,
        .model = model,
        .store_result = config.persist,
    });
    errdefer completion.deinit(allocator);

    var adapted = false;
    if (config.adapt_router) {
        modulator.update(router.analyzeSentiment(input));
        // Observe adaptive-weight persistence when a tracker is provided:
        // saveWeights' serialize buffer is alloc+freed internally (store.store
        // dupes it), so a TrackingAllocator records a balanced alloc/free without
        // touching any caller-owned buffer.
        var save_ta: ?core_memory.TrackingAllocator =
            if (config.tracker) |t| core_memory.TrackingAllocator.init(allocator, t) else null;
        const save_alloc = if (save_ta) |*ta| ta.allocator() else allocator;
        // A failed weight save must not discard the completion the caller owns;
        // surface it as "not adapted" rather than leaking or aborting.
        if (modulator.saveWeights(save_alloc, store)) |_| {
            adapted = true;
        } else |err| {
            std.log.warn("sea: router weight save failed: {s}", .{@errorName(err)});
            adapted = false;
        }
    }

    return .{
        .completion = completion,
        .evidence_count = evidence_count,
        .adapted = adapted,
        .query_task = plan.task,
    };
}

test "runLearnLoop tracks adaptive-weight persistence when a tracker is set" {
    if (!build_options.feat_wdbx or !build_options.feat_ai) return;
    const allocator = std.testing.allocator;

    var store = wdbx.Store.init(allocator);
    defer store.deinit();

    var tracker = core_memory.MemoryTracker.init(allocator);
    defer tracker.deinit();

    var result = try runLearnLoop(allocator, &store, "hello world", "abi-local", .{
        .persist = false,
        .adapt_router = true,
        .tracker = &tracker,
    });
    defer result.deinit(allocator);

    // The weight-save routed its serialize buffer through the tracker, so the
    // adaptive-weight persistence is now observable — and it was balanced
    // (alloc+free inside saveWeights), leaving the tracker net-zero.
    try std.testing.expect(result.adapted);
    try std.testing.expect(tracker.getTotalAllocated() > 0);
    try std.testing.expectEqual(tracker.getTotalAllocated(), tracker.getTotalFreed());
}

test "runLearnLoop persists a turn that a later related turn recalls as evidence" {
    if (!build_options.feat_wdbx or !build_options.feat_ai) return;
    const allocator = std.testing.allocator;

    var store = wdbx.Store.init(allocator);
    defer store.deinit();

    // First turn persisted (query/response vectors + completion metadata stored).
    var first = try runLearnLoop(allocator, &store, "the capital of france is paris", "abi-local", .{
        .persist = true,
        .adapt_router = false,
    });
    first.deinit(allocator);

    // A related second turn recalls the first turn as evidence (vectorCount>0 now).
    var second = try runLearnLoop(allocator, &store, "tell me about paris in france", "abi-local", .{
        .persist = true,
        .adapt_router = false,
    });
    defer second.deinit(allocator);

    try std.testing.expect(second.evidence_count > 0);
}

test "runLearnLoop surfaces the inferred task intent on the result" {
    if (!build_options.feat_wdbx or !build_options.feat_ai) return;
    const allocator = std.testing.allocator;

    var store = wdbx.Store.init(allocator);
    defer store.deinit();

    var result = try runLearnLoop(allocator, &store, "remember the prior decision we made", "abi-local", .{
        .persist = false,
        .adapt_router = false,
    });
    defer result.deinit(allocator);

    try std.testing.expectEqual(query_plan.TaskType.project_recall, result.query_task);
}

test "runLearnLoop uses saved adaptive weights for later learned completions" {
    if (!build_options.feat_wdbx or !build_options.feat_ai) return;
    const allocator = std.testing.allocator;

    var store = wdbx.Store.init(allocator);
    defer store.deinit();

    var modulator = router.AdaptiveModulator.deserialize("0.010000,0.980000,0.010000,8,0.050000");
    try modulator.saveWeights(allocator, &store);

    var result = try runLearnLoop(allocator, &store, "analyze the logical structure", "abi-local", .{
        .persist = false,
        .adapt_router = false,
    });
    defer result.deinit(allocator);

    try std.testing.expectEqual(ai.AgentProfile.aviva, result.completion.selected_profile);
    try std.testing.expect(std.mem.indexOf(u8, result.completion.output, "Aviva creative exploration") != null);
}

test {
    std.testing.refAllDecls(@This());
}

File: crates/abi-sea/src/learn_loop.rs
//! SEA self-learning loop: evidence → augment → adaptive complete → save weights.
//!
//! Ported from `src/features/sea/learn_loop.zig`.

use abi_ai::{
    AdaptiveModulator, CompletionResult, EMBED_DIM, EmptyInputError, MODULATOR_STORE_KEY,
    analyze_sentiment, complete_adaptive, completion, text_embedding,
};
use abi_wdbx::DurableStore;

use crate::evidence::{self, MAX_PROMPT_BYTES};
use crate::query_plan::{self, TaskType};

/// Configuration for one self-learning pass.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct LearnLoopConfig {
    /// Number of prior records to recall as evidence.
    pub evidence_limit: usize,
    /// Persist the completion (vectors + metadata + block) into the store.
    pub persist: bool,
    /// Update + save the adaptive persona-router weights from this turn.
    pub adapt_router: bool,
    /// Hard cap on the augmented prompt preamble (mirrors Zig's surface).
    pub max_prompt_bytes: usize,
}

impl Default for LearnLoopConfig {
    fn default() -> Self {
        Self {
            evidence_limit: 5,
            persist: true,
            adapt_router: true,
            max_prompt_bytes: MAX_PROMPT_BYTES,
        }
    }
}

/// Result of [`run_learn_loop`].
#[derive(Debug, Clone, PartialEq)]
pub struct LearnLoopResult {
    /// The underlying completion (persona response + audit).
    pub completion: CompletionResult,
    /// How many prior records were recalled.
    pub evidence_count: usize,
    /// Whether router weights were updated and saved.
    pub adapted: bool,
    /// Task intent inferred from the input.
    pub query_task: TaskType,
    /// Query / response vector ids and block hash, when `persist` succeeded.
    pub persisted: Option<PersistedIds>,
}

/// Ids produced when a completion is persisted.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PersistedIds {
    /// Stored query vector id.
    pub query_vector_id: u64,
    /// Stored response vector id.
    pub response_vector_id: u64,
    /// Audit-chain block hash as lowercase hex.
    pub block_hash_hex: String,
}

/// Run one SEA self-learning pass against `store`.
///
/// # Errors
///
/// Returns [`EmptyInputError`] when the generation input is empty.
pub fn run_learn_loop(
    store: &mut DurableStore,
    input: &str,
    model: &str,
    config: LearnLoopConfig,
    now_ms: i64,
) -> Result<LearnLoopResult, EmptyInputError> {
    let plan = query_plan::infer(input);
    let ctx = evidence::gather_evidence_with_plan(store, input, config.evidence_limit, &plan);
    let evidence_count = ctx.items.len();
    let augmented = evidence::augment_prompt(input, &ctx);
    let _ = config.max_prompt_bytes; // operative bound is evidence::MAX_PROMPT_BYTES

    // Generation sees the evidence-augmented prompt; routing uses the raw user
    // text so an explicit "Aviva, ..." address is not masked by the preamble.
    let persisted_weights = store.get(MODULATOR_STORE_KEY).map(str::to_string);
    let completion = complete_adaptive(&augmented, model, input, persisted_weights.as_deref())?;

    let mut adapted = false;
    if config.adapt_router {
        // Zig reloads weights from the store for the save path, independently
        // of the routing-time update inside complete_adaptive.
        let mut modulator = match store.get(MODULATOR_STORE_KEY) {
            Some(data) => AdaptiveModulator::deserialize(data),
            None => AdaptiveModulator::new(),
        };
        modulator.update(analyze_sentiment(input));
        let serialized = modulator.serialize();
        if store.put(MODULATOR_STORE_KEY, &serialized).is_ok() {
            adapted = true;
        }
    }

    // Zig's completeWithStoreAdaptive embeds request.input (the augmented
    // prompt) as the query vector — not the raw user text.
    let persisted = if config.persist {
        persist_completion(store, &augmented, &completion, now_ms)
    } else {
        None
    };

    Ok(LearnLoopResult {
        completion,
        evidence_count,
        adapted,
        query_task: plan.task,
        persisted,
    })
}

fn persist_completion(
    store: &mut DurableStore,
    generation_input: &str,
    result: &CompletionResult,
    now_ms: i64,
) -> Option<PersistedIds> {
    let query_vec: [f32; EMBED_DIM] = text_embedding(generation_input);
    let response_vec: [f32; EMBED_DIM] = text_embedding(&result.output);

    let query_id = store.put_vector(&query_vec).ok()?;
    let response_id = store.put_vector(&response_vec).ok()?;

    let metadata = completion::metadata_json(generation_input, result, query_id, response_id);
    let key = completion::metadata_key(query_id);
    store.put(&key, &metadata).ok()?;

    let block = store
        .add_block(
            result.selected_profile.label(),
            query_id,
            response_id,
            &metadata,
            now_ms,
        )
        .ok()?;

    Some(PersistedIds {
        query_vector_id: query_id,
        response_vector_id: response_id,
        block_hash_hex: block.hash.to_hex(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use abi_wdbx::StorePaths;
    use std::path::PathBuf;
    use std::sync::atomic::{AtomicU64, Ordering};

    static NEXT: AtomicU64 = AtomicU64::new(1);

    struct Scratch(PathBuf);
    impl Scratch {
        fn new() -> Self {
            let n = NEXT.fetch_add(1, Ordering::Relaxed);
            let path =
                std::env::temp_dir().join(format!("abi_sea_learn_{}_{n}", std::process::id()));
            std::fs::create_dir_all(&path).unwrap();
            Self(path)
        }
    }
    impl Drop for Scratch {
        fn drop(&mut self) {
            let _ = std::fs::remove_dir_all(&self.0);
        }
    }

    fn open() -> (Scratch, DurableStore) {
        let dir = Scratch::new();
        let store = DurableStore::open(StorePaths::new(&dir.0)).unwrap();
        (dir, store)
    }

    #[test]
    fn surfaces_inferred_task_intent() {
        let (_dir, mut store) = open();
        let result = run_learn_loop(
            &mut store,
            "remember the prior decision we made",
            "abi-local",
            LearnLoopConfig {
                persist: false,
                adapt_router: false,
                ..LearnLoopConfig::default()
            },
            1_700_000_000_000,
        )
        .unwrap();
        assert_eq!(result.query_task, TaskType::ProjectRecall);
        assert!(!result.adapted);
        assert!(result.persisted.is_none());
    }

    #[test]
    fn adapts_and_persists_weights() {
        let (_dir, mut store) = open();
        let result = run_learn_loop(
            &mut store,
            "hello world",
            "abi-local",
            LearnLoopConfig {
                persist: false,
                adapt_router: true,
                ..LearnLoopConfig::default()
            },
            1_700_000_000_000,
        )
        .unwrap();
        assert!(result.adapted);
        assert!(store.get(MODULATOR_STORE_KEY).is_some());
    }

    #[test]
    fn uses_saved_adaptive_weights_for_later_completions() {
        let (_dir, mut store) = open();
        // Heavy aviva prior — same shape as Zig's learn_loop test.
        store
            .put(MODULATOR_STORE_KEY, "0.010000,0.980000,0.010000,8,0.050000")
            .unwrap();
        let result = run_learn_loop(
            &mut store,
            "analyze the logical structure",
            "abi-local",
            LearnLoopConfig {
                persist: false,
                adapt_router: false,
                ..LearnLoopConfig::default()
            },
            1_700_000_000_000,
        )
        .unwrap();
        assert_eq!(
            result.completion.selected_profile,
            abi_ai::AgentProfile::Aviva
        );
        assert!(result.completion.output.contains("Aviva direct expert"));
    }

    #[test]
    fn first_turn_is_recalled_as_evidence_on_related_second_turn() {
        let (_dir, mut store) = open();
        let first = run_learn_loop(
            &mut store,
            "the capital of france is paris",
            "abi-local",
            LearnLoopConfig {
                persist: true,
                adapt_router: false,
                ..LearnLoopConfig::default()
            },
            1_700_000_000_000,
        )
        .unwrap();
        assert!(first.persisted.is_some());

        let second = run_learn_loop(
            &mut store,
            "tell me about paris in france",
            "abi-local",
            LearnLoopConfig {
                persist: true,
                adapt_router: false,
                ..LearnLoopConfig::default()
            },
            1_700_000_000_001,
        )
        .unwrap();
        assert!(second.evidence_count > 0);
    }

    #[test]
    fn corrupted_weights_route_like_missing_weights() {
        let input = "analyze the logical structure";
        let (_d1, mut empty) = open();
        let baseline = run_learn_loop(
            &mut empty,
            input,
            "abi-local",
            LearnLoopConfig {
                persist: false,
                adapt_router: false,
                ..LearnLoopConfig::default()
            },
            1,
        )
        .unwrap();

        let (_d2, mut corrupt) = open();
        corrupt
            .put(MODULATOR_STORE_KEY, "nan,0.3,0.4,1,0.3")
            .unwrap();
        let corrupted = run_learn_loop(
            &mut corrupt,
            input,
            "abi-local",
            LearnLoopConfig {
                persist: false,
                adapt_router: false,
                ..LearnLoopConfig::default()
            },
            1,
        )
        .unwrap();

        assert_eq!(
            baseline.completion.selected_profile,
            corrupted.completion.selected_profile
        );
        assert_eq!(baseline.completion.output, corrupted.completion.output);
    }
}
